text/3698-declarative-derive-macros.md

```diff
@@ -0,0 +1,122 @@
+
+# Motivation
+[motivation]: #motivation
+
+Many crates support deriving their traits with `derive(Trait)`. Today, this
+requires defining proc macros, in a separate crate, typically with several
+additional dependencies adding substantial compilation time, and typically
+guarded by a feature that users need to remember to enable.
+
+However, many common cases of derives don't require any more power than an
+ordinary `macro_rules!` macro. Supporting these common cases would allow many
+crates to avoid defining proc macros, reduce dependencies and compilation time,
+and provide these macros unconditionally without requiring the user to enable a
+feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can define a macro to implement `derive(MyTrait)` by defining a
+`macro_rules!` macro with the `#[macro_derive]` attribute. Such a macro can
+create new items based on a struct, enum, or union. Note that the macro can
+only append new items; it cannot modify the item it was applied to.
+
+For example:
+
+```rust
+trait Answer { fn answer(&self) -> u32; }
+
+#[macro_derive]
+macro_rules! Answer {
+    // Simplified for this example
+    (struct $n:ident $_:tt) => {
+        impl Answer for $n {
+            fn answer(&self) -> u32 { 42 }
+        }
+    };
+}
+
+#[derive(Answer)]
+struct Struct;
+
+fn main() {
+    let s = Struct;
+    assert_eq!(42, s.answer());
+}
+```
+
+Derive macros defined using `macro_rules!` follow the same scoping rules as
+any other macro, and may be invoked by any path that resolves to them.
+
+A derive macro may share the same path as a trait of the same name. For
+instance, the name `mycrate::MyTrait` can refer to both the `MyTrait` trait and
+the macro for `derive(MyTrait)`. This is consistent with existing derive
+macros.
+
+A derive macro may also define *helper attributes*. These attributes are
+[inert](https://doc.rust-lang.org/reference/attributes.html#active-and-inert-attributes),
+and their only purpose is to be fed into the derive macro that defined them.
+That said, they can be seen by all macros.
+
+To define helper attributes, put an attributes key in the `macro_derive`
+attribute, with a comma-separated list of identifiers for helper attributes:
+`#[macro_derive(attributes(helper))]`. The derive macro can process the
+`#[helper]` attribute, along with any arguments to it, as part of the item the
+derive macro was applied to.
+
+If a derive macro mistakenly emits the token stream it was applied to
+(resulting in a duplicate item definition), the error the compiler emits for
+the duplicate item should hint to the user that the macro was defined
+incorrectly, and remind the user that derive macros only append new items.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Adding this feature will allow many crates in the ecosystem to drop their proc
+macro crates and corresponding dependencies, and decrease their build times.
```

Comment:
I'm assuming most cases will need tt-munching which I thought is very slow for compiler performance, so unsure if this will be a significant win.

Reply:
@epage I would not expect this to need full tt-munching in many cases. And regardless, I'd expect a bit of tt-munching to still be faster than building the additional dependencies used by most proc macros.

Reply:
I don't know if its because I've found iteration harder than recursion in macro_rules but my gut usually leans towards "if its non-trivial, it'll use tt-munching". 

To check on that intuition, I'm looking at [bitflags](https://crates.io/crates/bitflags) for imagining it being a derive.
- The top-level macro uses recursion to support multiple structs in a `bitflags! {}`, which wouldn't be needed
  - https://github.com/bitflags/bitflags/blob/a59b364a7e9dbe5672f1e278b0e2a2ae426acbad/src/lib.rs#L457
- Processing of attributes
  - Performance would scale with the number of attributes or inner attributes.  Going to an extreme in complexity, I only expect 3-4 inner attributes for clap, 10 at most.  So this is likely not to be a problem
  - https://github.com/bitflags/bitflags/blob/a59b364a7e9dbe5672f1e278b0e2a2ae426acbad/src/lib.rs#L795-L807

So maybe not as bad as I expected.

> And regardless, I'd expect a bit of tt-munching to still be faster than building the additional dependencies used by most proc macros.

For full builds, likely yes.  For incremental builds, tt-munching is likely to be worse and likely harder to predict when the macro expansion caching would be worth it compared to proc-macros.

Reply:
I think my intuition agrees with what your investigations surfaced: I would expect simple case like "look at each struct" (not needed) or "look at each field" or "don't need to touch any fields" to generally not need a muncher, while more complex cases like parsing out attributes are more likely to need a muncher-style design.

Reply:
> For full builds, likely yes. For incremental builds, tt-munching is likely to be worse and likely harder to predict when the macro expansion caching would be worth it compared to proc-macros.

Fair enough. That seems worth investigating once we have an implementation. I am hoping that caching solves that problem, such that you only have to re-run the macro on an incremental build in which the ADT definition *changes*.